README.md
# pytest-pyppeteer
A plugin to run [pyppeteer](https://github.com/pyppeteer/pyppeteer) in pytest.

![PyPI - Python Version](https://img.shields.io/pypi/pyversions/pytest-pyppeteer)
[![GitHub issues](https://img.shields.io/github/issues-raw/luizyao/pytest-pyppeteer)](https://github.com/luizyao/pytest-pyppeteer/issues)
[![PyPI](https://img.shields.io/pypi/v/pytest-pyppeteer)](https://pypi.org/project/pytest-pyppeteer/)
[![Downloads](https://pepy.tech/badge/pytest-pyppeteer)](https://pepy.tech/project/pytest-pyppeteer)
[![Code style: black](https://img.shields.io/badge/code%20style-black-000000.svg)](https://github.com/psf/black)
[![Documentation Status](https://readthedocs.org/projects/pytest-pyppeteer/badge/?version=latest)](https://pytest-pyppeteer.readthedocs.io/en/latest/?badge=latest)
[![中文博客](https://img.shields.io/badge/blog-%E4%B8%AD%E6%96%87%E5%8D%9A%E5%AE%A2-yellowgreen)](https://luizyao.com/blog/post/2020-11-06-run-pyppeteer-in-pytest/)

# Installation
You can install pytest-pyppeteer via [pip](https://pypi.org/project/pip/):

```bash
$ pip install pytest-pyppeteer
```

or install the latest one on Github:

```bash
pip install git+https://github.com/luizyao/pytest-pyppeteer.git
```


# Quickstart
For example, **The Shawshank Redemption** deserves a 9.0 or higher rating on [douban.com](https://movie.douban.com).

```python
from dataclasses import dataclass

from pytest_pyppeteer.models import Browser


@dataclass(init=False)
class Elements:
    """Collect locators of page objects, no matter XPath or CSS Selector."""

    # query input
    query = "#inp-query"

    # search button
    apply = ".inp-btn > input:nth-child(1)"

    # the first result
    first_result = (
        "#root > div > div > div > div > div:nth-child(1) > div.item-root a.cover-link"
    )

    # rating
    rating = (
        "#interest_sectl > div.rating_wrap.clearbox > div.rating_self.clearfix > strong"
    )


async def test_lifetimes(pyppeteer: Browser):
    page = await pyppeteer.new_page()
    await page.goto('https://movie.douban.com/')

    await page.type(Elements.query, "The Shawshank Redemption")
    await page.click(Elements.apply)

    await page.waitfor(Elements.first_result)
    await page.click(Elements.first_result)

    await page.waitfor(Elements.rating)
    rating = await page.get_value(Elements.rating)
    
    assert float(rating) >= 9.0
```

![quickstart](docs/image/quickstart.gif)

# Usage

## Command line options

### `--executable-path`

You can specify the path to a Chromium or Chrome executable. otherwise pytest-pyppeteer will use the default installation location of Chrome in current platform, but now only support `win64`, `win32` and `mac` platform.

For other platforms, pyppeteer will downloads the recent version of Chromium when called first time. If you don’t prefer this behavior, you can specify an exact path by override this fixture:

```python
@pytest.fixture(scope="session")
def executable_path(executable_path):
    if executable_path is None:
        return "path/to/Chrome/or/Chromium"
    return executable_path
```

> **Note**:
>
> The default installation location of Chrome in different platform:
>
> - `win64`: C:/Program Files/Google/Chrome/Application/chrome.exe
> - `win32`: C:/Program Files (x86)/Google/Chrome/Application/chrome.exe
> - `mac`: /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

### `--headless`

Run browser in headless mode.

### `--args`

Additional args to pass to the browser instance.

For example, specify a proxy:

```bash
$ pytest --args proxy-server "localhost:5555,direct://" --args proxy-bypass-list "192.0.0.1/8;10.0.0.1/8"
```

Or by override the `args` fixture:

```python
@pytest.fixture(scope="session")
def args(args) -> List[str]:
    return args + [
        "--proxy-server=localhost:5555,direct://",
        "--proxy-bypass-list=192.0.0.1/8;10.0.0.1/8",
    ]
```

### `--window-size`

The default browser size is 800*600, you can use this option to change this behavior:

```bash
$ pytest --window-size 1200 800
```

`--window-size 0 0` means to starts the browser maximized.

### `--slow`

Slow down the pyppeteer operate in milliseconds. Defaults to `0.0`.

## No matter selector or xpath

`pyppeteer` fixture provide a `pytest_pyppeteer.models.Browser` instance, its usage is almost the same as `pyppeteer.browser.Browser`, except that it provides a new instance method: `new_page()`, which is similar to `newPage()`, but it returns a `pytest_pyppeteer.models.Page` instead of `pyppeteer.page.Page`.

`pytest_pyppeteer.models.Page`’s usage is also the same as `pyppeteer.page.Page`, but it provides some new instance methods, and override some methods. For example, you can query an element by selector or xpath in just same method `query_locator` instead of original `querySelector` and `xpath`.

You can also get an original `Page` by `pyppeteer.newPage()`.

## `options` marker

You can override some command line options in the specified test.

For example, auto-open a DevTools panel:

```python
import asyncio

import pytest


@pytest.mark.options(devtools=True)
async def test_marker(pyppeteer):
    await pyppeteer.new_page()
    await asyncio.sleep(2)
```

![options marker](docs/image/options_marker.gif)

# Advanced Usage

## Control multiple browsers asynchronously

You can easily to control multiple browsers at the same time.

For example, query the **The Shawshank Redemption**’s movie and book rating on [douban.com](https://movie.douban.com/) at the same time, then compare them.

```python
import asyncio
from dataclasses import dataclass
from typing import TYPE_CHECKING, Callable

import pytest

if TYPE_CHECKING:
    from .models import Browser, Page


@dataclass
class Elements:
    query = "#inp-query"
    apply = ".inp-btn > input:nth-child(1)"


@dataclass
class BookElements(Elements):
    url = "https://book.douban.com/"

    result = '(//*[@class="item-root"])[1]/a'
    rating = "#interest_sectl > div > div.rating_self.clearfix > strong"


@dataclass
class MovieElements(Elements):
    url = "https://movie.douban.com/"

    result = (
        "#root > div > div > div > div > div:nth-child(1) > div.item-root a.cover-link"
    )
    rating = (
        "#interest_sectl > div.rating_wrap.clearbox > div.rating_self.clearfix > strong"
    )


async def query_rating(pyppeteer: "Browser", name: str, elements: "Elements"):
    page: Page = await pyppeteer.new_page()

    await page.goto(elements.url)

    await page.type(elements.query, name)
    await page.click(elements.apply)

    await page.waitfor(elements.result)
    await page.click(elements.result)

    await page.waitfor(elements.rating)
    rating = await page.get_value(elements.rating)
    return rating


async def test_multiple_browsers(pyppeteer_factory: "Callable"):
    pyppeteer1 = await pyppeteer_factory()
    pyppeteer2 = await pyppeteer_factory()

    movie_rating, book_rating = await asyncio.gather(
        query_rating(pyppeteer1, "The Shawshank Redemption", MovieElements),
        query_rating(pyppeteer2, "The Shawshank Redemption", BookElements),
    )

    assert movie_rating == book_rating
```

![multiple_browsers](docs/image/multiple_browsers.gif)

# License

Distributed under the terms of the [MIT](http://opensource.org/licenses/MIT) license, pytest-pyppeteer is free and open source software.


# Issues
If you encounter any problems, please [file an issue](https://github.com/luizyao/pytest-pyppeteer/issues) along with a detailed description.


CHANGELOG.md
# Changelog

<a name="0.2.5"></a>
## 0.2.5 (2021-02-16)

### Added

- ✅ update tests [[7c0c852](https://github.com/luizyao/pytest-pyppeteer/commit/7c0c852004143275d3a26fc34b34ebf74760fa9e)]

### Changed

- ⬆️ upgrade pyppeteer: 0.2.3 &#x3D;&gt; 0.2.5 [[aaef028](https://github.com/luizyao/pytest-pyppeteer/commit/aaef0286be780c2bf7697dcd87a7eddf36f1eba2)]

### Fixed

- ✏️ add chinese blog badge [[46dd76d](https://github.com/luizyao/pytest-pyppeteer/commit/46dd76d98c027a4590778bed346ba164a42fe062)]

### Security

- 🔒 update lxml version.(&gt;&#x3D; 4.6.2) ([#21](https://github.com/luizyao/pytest-pyppeteer/issues/21)) [[3788590](https://github.com/luizyao/pytest-pyppeteer/commit/3788590bd100310b4057f1f452462b1a35fd36ab)]

### Miscellaneous

- 💡 update docstring for &quot;--slow&quot; [[1464fa6](https://github.com/luizyao/pytest-pyppeteer/commit/1464fa638656be10d6f8554f4d9656afac06022b)]


<a name="0.2.3"></a>
## 0.2.3 (2020-11-26)

### Added

- ✨ add &quot;--slow&quot; command option [63d641b]
- ✅ update test and documents [d21e475]

### Removed

- ➖ Remove gitee pages [6453d11]

### Fixed

- ✏️ fix typos in the documents [e75932a]
- 🐛 fix issue #19 [98852f9]

### Miscellaneous

- 💩 change the type of Chrome executable from &quot;str&quot; to &quot;Path&quot; [cd3838f]
- 📝 add changelog file [b6531e3]
- 📝 update documents [0d13485]
- 📝 remove invalid shields link [f92d955]
- 🔨 add edit on github link in the documents [e73180b]


<a name="0.2.1"></a>
## 0.2.1 (2020-10-16)

### Added

- ✨ add github action [d31e467]

### Fixed

- ✏️ fix syntax typos [adcc356]
- 🐛 fix issue #18 [b1b2246]

### Miscellaneous

- 📝 update documents link in readme [b79265f]
- 🔨 use githubpages sphinx extension [4f70476]
- 🔨 add .nojekyll file for github page [7852599]
- 📝 update documents [d9210f2]


<a name="0.2.0"></a>
## 0.2.0 (2020-10-02)

### Added

- ✅ add tests [bee660b]
- ✨ support to start the browser maximized [f632ea5]
- ✨ add new hook [bd04729]
- ✨ add options marker [989150e]
- ✨ add errors module [37dece5]
- ✨ add &quot;--executable-path&quot; command line option [664dd7c]
- ✨ integrate with pytest-asyncio plugin [0ac2158]
- 🎉 initial commit [345dc86]

### Changed

- 🎨 format imports [d0f81ab]
- 🎨 format imports in docs [b8eff72]
- 🎨 format imports [705bf95]
- 🔧 add readthedocs configure file [69f46ab]
- ⚡ add the fault tolerance processing [b44ce68]
- 🔧 use pipenv to manage environemnt [0f6f230]

### Removed

- 🔥 remove requirements.txt [ce7a3df]

### Fixed

- 🐛 fix unprintable error class [d57af5a]
- 🐛 fix different &quot;--window-size 0 0&quot; behaviors between mac and windows [ff30cdd]
- 🐛 import missing module [ca41ba5]
- ✏️ fix import error [fd668ba]

### Miscellaneous

- 📝 fix typos in readme [95cfc6e]
- 📝 update docs [e6569ac]
- 📝 add docs [5065798]
- 💡 fix format error in docstring [fa61658]
- 🔨 remove unsupported python version [b5190cf]
- 💡 update reference in the comment [242a9a4]
- 💡 fix docstring format error [5f6f9f7]
- 🚩 add Browser and Page model [4b033c7]
- 🚧 add viewport model [b00da6f]
- 💡 update &quot;Option&quot; model docstring [ea0ac55]
- 🚧 add pyppeteer options model [662b832]
- 💡 modify docstring [1c7aa0d]
- 🔨 remove unsupported python version [b62fa64]
- 🔨 remove unsupported python version [d68228f]
- 🔨 remove invalid chars [f0933f5]




src/pytest_pyppeteer/plugin.py
import asyncio
import inspect
import logging
from typing import TYPE_CHECKING, Any, Awaitable, Callable, List, Optional

import pytest
from pyppeteer import launch

from pytest_pyppeteer.models import Browser, Options
from pytest_pyppeteer.utils import CHROME_EXECUTABLE, current_platform

if TYPE_CHECKING:
    from _pytest.config import Config, PytestPluginManager
    from _pytest.config.argparsing import Parser
    from _pytest.fixtures import FixtureRequest
    from _pytest.nodes import Item
    from _pytest.runner import TestReport

LOGGER = logging.getLogger(__name__)


def pytest_addhooks(pluginmanager: "PytestPluginManager") -> None:
    """Add new hooks.

    :param _pytest.config.PytestPluginManager pluginmanager:
    :return: ``None``
    """
    from . import hooks

    pluginmanager.add_hookspecs(hooks)


def pytest_configure(config: "Config") -> None:
    """Perform initial configuration as follows:

    * Add ``pytest.mark.option`` marker to the ini-file option.

    :param _pyest.config.Config config: pytest config object.
    :return: None

    .. note::

        This is a ``pytest hook function`` which is called for
        every plugin and initial conftest file after command
        line options have been parsed. After that, the hook
        is called for other conftest files as they are imported.
    """
    config.addinivalue_line(
        "markers",
        "options(kwargs): add or change existing options. "
        "specify options as keyword argument, e.g. "
        "@pytest.mark.options(devtools=True)",
    )


@pytest.mark.tryfirst
def pytest_collection_modifyitems(items: List["Item"]) -> None:
    """Modify the items collected by pytest as follows:

    * Because all test items using pyppeteer should be an
      asyncio coroutine, here the `pytest.mark.asyncio`_
      marker is automatically added to each test item
      collected by pytest.

    :param List[pytest.Item] items: the list of items objects
           collected by pytest.
    :return: None

    .. note::

        This is a ``pytest hook function`` which is called
        after collection of all test items is completed.

    ..  _pytest.mark.asyncio:
        https://github.com/pytest-dev/pytest-asyncio#pytestmarkasyncio
    """
    items[:] = [add_asyncio_marker(item) for item in items]


def add_asyncio_marker(item: "Item") -> "Item":
    """Add `pytest.mark.asyncio`_ marker to the specified item.

    If the marker is already exists, return the item directly.

    :param pytest.Item item: the pytest item object.
    :return: the marked item object.

    ..  _pytest.mark.asyncio:
        https://github.com/pytest-dev/pytest-asyncio#pytestmarkasyncio
    """
    if "asyncio" not in item.keywords and is_coroutine(item.obj):
        item.add_marker(pytest.mark.asyncio)
    return item


def is_coroutine(obj: Any) -> bool:
    """Check to see if an object is really an asyncio coroutine.

    :param Any obj: any object.
    :return: `True` or `False`.
    """
    return asyncio.iscoroutinefunction(obj) or inspect.isgeneratorfunction(obj)


@pytest.hookimpl(hookwrapper=True, tryfirst=True)
def pytest_runtest_makereport(item: "Item") -> None:
    """Implement this pytest hook in wrapper mode, the added
    behaviors as follows:

    * Register a new
      :py:func:`hooks.pytest_pyppeteer_runtest_makereport_call_debug`
      hook which called when a actual failing test calls not setup/teardown.

    :param pytest.Item item: the pytest item object.
    :return: None
    """
    # execute all other hooks to obtain the report object
    outcome = yield
    res: TestReport = outcome.get_result()
    setattr(item, f"res_{res.when}", res)

    # we only deal with actual failing test calls, not setup/teardown
    if res.when == "call" and res.failed:
        hook: List[
            Awaitable
        ] = item.ihook.pytest_pyppeteer_runtest_makereport_call_debug(item=item)
        if hook:
            loop = asyncio.get_event_loop()
            loop.run_until_complete(hook[0])


def pytest_addoption(parser: "Parser") -> None:
    """Register new command line arguments and ini-file values.

    Create a new command line option group named `pyppeteer`, and add the
    following new options in it:

    * ``--executable-path``: path to a Chromium or Chrome executable.

    * ``--headless``: run browser in headless mode.

    * ``--args``: additional args to pass to the browser instance. more
      details refer to :py:func:`args` fixture.

    * ``--window-size``: set the initial browser window size. Defaults
      to 800 * 600. ``--window-size 0 0`` means to starts the browser
      maximized.

    * ``--slow``: slow down the pyppeteer operate in milliseconds.
      Defaults to ``0.0``.

    :param _pytest.config.argparsing.Parser parser: parser for command
           line arguments and ini-file values.
    :return: None

    .. note::

        This is a ``pytest hook function`` which be called once at the beginning
        of a test run to register argparse-style options and ini-style config values.

        There are two ways to register new options, respectively:

        * To register a command line option, call
          :py:meth:`parser.addoption(...) <pytest:_pytest.config.argparsing.Parser.addoption>`.

        * To register an ini-file option, call
          :py:meth:`parser.addini(...) <pytest:_pytest.config.argparsing.Parser.addini>`.

        And the options can later be accessed through the
        :py:class:`Config <pytest:_pytest.config.Config>` object, respectively:

        * To retrieve the value of a command line option, call
          :py:meth:`config.getoption(name) <pytest:_pytest.config.Config.getoption>`.

        * To retrieve a value read from an ini-style file, call
          :py:meth:`config.getini(name) <pytest:_pytest.config.Config.getini>`.

        The ``Config`` object is passed around on many pytest internal objects via the
        ``.config`` attribute or can be retrieved as the
        :py:func:`pytestconfig <pytest:_pytest.fixtures.pytestconfig>` fixture.
    """
    # Create a option group named "pyppeteer"
    group = parser.getgroup("pyppeteer", description="pyppeteer")
    group.addoption(
        "--executable-path", help="path to a Chromium or Chrome executable."
    )

    group.addoption(
        "--headless", action="store_true", help="run browser in headless mode."
    )

    group.addoption(
        "--args",
        action="append",
        nargs="+",
        help="additional args to pass to the browser instance.",
    )

    group.addoption(
        "--window-size",
        nargs=2,
        default=["800", "600"],
        help="set the initial browser window size.",
    )

    group.addoption(
        "--slow", type=float, default=0.0, help="slow down the operate in milliseconds."
    )


@pytest.fixture(scope="session")
def executable_path(pytestconfig: "Config") -> Optional[str]:
    """``Session-scoped fixture`` that return Chrome or Chromium executable path.

    The fixture behaviors follow this procedure:

    1. Return the value passed in from command line option of `--executable-path`,
       if it's not ``None``.

    2. Return the default installation location of Chrome in current platform,
       but now only support ``win64``, ``win32`` and ``mac`` platform.

       For other platforms, pyppeteer will downloads the recent version of Chromium
       when called first time. If you don't prefer this behavior, you can specify
       an exact path by overwrite this fixture:

       Example::

           @pytest.fixture(scope="session")
           def executable_path(executable_path):
               if executable_path is None:
                   return "path/to/Chrome/or/Chromium"
                   return executable_path


    :param _pytest.config.Config pytestconfig: a session-scoped fixture that return
           config object.
    :return: return Chrome or Chromium executable path string. but if current platform
             isn't supported, return ``None``.
    """
    path = pytestconfig.getoption("--executable-path")
    if path:
        return path

    try:
        return CHROME_EXECUTABLE[current_platform()]
    except OSError as e:
        # Unsupported platform
        LOGGER.error(e)
        return None


@pytest.fixture(scope="session")
def args(pytestconfig: "Config") -> List[str]:
    """``Session-scoped fixture`` that return a list of additional
    args in the `List of Chromium Command Line Arguments`_ to pass
    to the browser instance.

    You can use it by command-line option:

    Example::

        $ pytest --args proxy-server "localhost:5555,direct://" --args proxy-bypass-list "192.0.0.1/8;10.0.0.1/8"

    Or overwrite it in your test:

    Example::

        @pytest.fixture(scope="session")
        def args(args) -> List[str]:
            return args + [
                "--proxy-server=localhost:5555,direct://",
                "--proxy-bypass-list=192.0.0.1/8;10.0.0.1/8",
            ]

    :param _pytest.config.Config pytestconfig: a session-scoped
           fixture that return config object.
    :return: a list of arguments string. return ``list()`` if no
             ``--args`` passed in the command-line.

    .. _List of Chromium Command Line Arguments:
       https://peter.sh/experiments/chromium-command-line-switches/
    """
    args: list = pytestconfig.getoption("--args")
    if args:
        return ["--" + "=".join(arg_list) for arg_list in args]
    else:
        return list()


@pytest.fixture(scope="session")
def session_options(
    pytestconfig: "Config", args: List[str], executable_path: str
) -> "Options":
    """``Session-scoped fixture`` that return a
    :py:class:`models.Options` object used to initialize browser.

    :param _pytest.config.Config pytestconfig: a session-scoped
           fixture that return config object.
    :param List[str] args: a session-scoped fixture that return
           a list of additional args to pass to the browser instance.
    :param str executable_path: a session-scoped fixture that return
           Chrome or Chromium executable path.
    :return: a :py:class:`models.Options` object used to initialize
             browser.
    """
    headless: bool = pytestconfig.getoption("--headless")
    width, hight = pytestconfig.getoption("--window-size")
    slow: float = pytestconfig.getoption("--slow")
    if width == "0" and hight == "0":
        if current_platform() == "mac":
            args.append("--start-fullscreen")
        else:
            args.append("--start-maximized")
    else:
        args.append("--window-size={},{}".format(width, hight))
    return Options(
        args=args, executablePath=executable_path, headless=headless, slowMo=slow
    )


@pytest.fixture
def options(request: "FixtureRequest", session_options: "Options") -> "Options":
    """``Function-scoped fixture`` that return a
    :py:class:`models.Options` object used to initialize browser.

    This fixture contains all of :py:func:`session_options`, plus
    any options specified by the options markers. Any change to
    these options will apply only to the tests covered by scope
    of the fixture override.

    Example::

        @pytest.mark.options(devtools=True)
        async def test_options_mark(pyppeteer):
            ...

    :param _pytest.fixture.FixtureRequest request: A request
           object gives access to the requesting test context.
    :param Options session_options: a :py:class:`models.Options`
           object from :py:func:`session_options`.
    :return: a :py:class:`models.Options` object used to initialize
             browser.
    """
    return session_options.copy(update=get_options_from_markers(request.node))


def get_options_from_markers(item: "Item") -> dict:
    """Get the options from the ``options`` markers of test item.
    And there are only apply on the current test item.

    :param Item item: the test item object.
    :return: an dict contains options.
    """
    options_dict = dict()
    for node, mark in item.iter_markers_with_node("options"):
        LOGGER.debug(
            "{0} marker <{1.name}> "
            "contains kwargs <{1.kwargs}>".format(node.__class__.__name__, mark)
        )
        options_dict.update(mark.kwargs)
    LOGGER.info("Options from markers: {}".format(options_dict))
    return options_dict


@pytest.fixture
async def pyppeteer_factory(options: "Options") -> Callable:
    """``Function-scoped fixture`` that return a pyppeteer
    browser factory.

    :param Options options: a :py:class:`models.Options`
           object used to initialize browser
    :yield: a pyppeteer browser factory.
    """
    browsers: List[Browser] = list()

    async def _factory() -> "Browser":
        LOGGER.info("Options to initialize browser instance: {}".format(options))
        b = Browser(pyppeteer_browser=await launch(options=options.dict()))
        browsers.append(b)
        return b

    yield _factory

    for browser in browsers:
        await browser.close()


@pytest.fixture
async def pyppeteer(pyppeteer_factory: "Callable") -> "Browser":
    """``Function-scoped fixture`` that return a pyppeteer
    browser instance.

    :param Callable pyppeteer_factory: pyppeteer factory
    :yield: a pyppeteer browser instance.
    """
    yield await pyppeteer_factory()
